cryptographic_estimators/MREstimator/MRAlgorithms/support_minors.py

```diff
@@ -0,0 +1,261 @@
+            4
+        """
+        return self._get_optimal_parameter(MR_REDUCED_NUMBER_OF_COLUMNS)
+
+    @optimal_parameter
+    def variant(self):
+        """
+        Return the optimal `variant`
+
+        EXAMPLES::
+
+            sage: from cryptographic_estimators.MREstimator.MRAlgorithms.support_minors import SupportMinors
+            sage: from cryptographic_estimators.MREstimator.mr_problem import MRProblem
+            sage: SM = SupportMinors(MRProblem(q=7, m=9, n=10, k=15, r=4))
+            sage: SM.nprime()
+            'strassen'
+        """
+        return self._get_optimal_parameter(MR_VARIANT)
+
+
+    def _expected_dimension_of_support_minors_equations(self,q, m, n, K, r, b):
+        """
+        Return the expected number of linearly independent support minors equations
+
+        """
+        if q == 2:
+            temp = 0
+            for j in range(1, b + 1):
+                temp += sum([(-1) ** (i + 1) * binomial(n, r + i) * binomial(m + i - 1, i) * binomial(K, j - i) for i in
+                             range(1, j + 1)])
+        else:
+            temp = sum(
+                [(-1) ** (i + 1) * binomial(n, r + i) * binomial(m + i - 1, i) * binomial(K + b - i - 1, b - i) for i in
+                 range(1, b + 1)])
+        return temp
+
+
+    def _dimension(self,q, n, K, r, b):
+        """
+        Dimension of the smallest vector space spanned by monomials containing the support minors equations
+        """
+        if q == 2:
+            temp = binomial(n, r) * sum([binomial(K, j) for j in range(1, b + 1)])
+        else:
+            temp = binomial(n, r) * binomial(K + b - 1, b)
+        return temp
+
+
+    def _are_parameters_invalid(self, parameters: dict):
+        """
+        Specifies constraints on the parameters
+        """
+        a = parameters["a"]
+        lv = parameters["lv"]
+        b = parameters["b"]
+        nprime = parameters["nprime"]
+        q, m, _, k_reduced, r =  self.get_problem_parameters_reduced(a, lv)
+        dim = self._dimension(q, nprime, k_reduced +  1, r, b) - 1
+        exp = self._expected_dimension_of_support_minors_equations(q, m, nprime, k_reduced + 1, r, b)
+        return dim < exp
+
+    def _valid_choices(self):
+        """
+        Generator which yields on each call a new set of valid parameters based on the `_parameter_ranges` and already
+        set parameters in `_optimal_parameters`
+
+        """
+        new_ranges = self._fix_ranges_for_already_set_parameters()
+        _ = new_ranges.pop(MR_VARIANT)
+        indices = {i: new_ranges[i]["min"] for i in new_ranges}
+        variant = MR_STRASSEN
+        keys = [i for i in indices]
+        stop = False
+        while not stop:
+            if not self._are_parameters_invalid(indices):
+                aux = indices.copy()
+                aux.update({MR_VARIANT: variant})
+                yield aux
+            indices[next(iter(indices))] += 1
+            for i in range(len(keys)):
```

Comment:
Thats seems hacky: As far as I understand you need to override to original `_valid_choices` to be able to change the matrix algorithm from Strassen to Block_Wiedemann. I think it would far easier to define these two types as an enumeration:
```python
from enum import Enum

class MR_VARIANT(Enum):
     MR_STRASSEN = 1
     MR_BLOCK_WIEDEMANN = 2
```
Now you can set the limits in the constructor as:
```
self.set_parameter_ranges('variant', 1, 2)
```

And with this you should be able to delete the whole function

Reply:
I agree, unless there is a misunderstanding why this function was overidden.

Reply:
Since there are many following distinctions made based on this, I am wondering if it would make sense to just create two algorithms for this -- SM using Block_Wiedemann and SM using Strassen. Or is one always superior to the other for a given set (in terms of time and memory)?

Reply:
@FloydZ  I fixed it. The only problem with this approach is that in the printed table you see `'variant': 1` instead of the more clear one `variant = strassen`

Reply:
extend the `MR_VARIANT` class with:
```python
def __str__(self):
   return self.name
```
and the name should be printed